### DISCUSSION TITLE
feature: add glossy tabs widget like charmbracelet

### DISCUSSION THREAD

**Discussion by crocuda** [ratatui/ratatui#1364] [Ideas]
## Problem

Tabs are cool but lack border style configurations.

## Solution

Can I update the Tabs widget with titles that can render like the following?

 ```text
 ╭──────╮╭──────╮╭──────╮╭──────╮
 │ Tab1 ││ Tab2 ││ Tab3 ││ Tab4 │
 │      └┴──────┴┴──────┴┴──────┴
 ```
 ```text
 ╭──────╮──────╮──────╮──────╮
 │ Tab1 │ Tab2 │ Tab3 │ Tab4 │
 │      └──────┴──────┴──────┴
 ```

This will have to be another widget because the Tabs Struct misses some fields like the ones on the Block widgets for such customization.

**Comment by joshka** [comment]
I like the idea of making the widgets generally look pretty. I'd like to suggest looking at this holistically.

What would this look like if you designed an API based on the existing Tabs widget? Perhaps when looking at this you may find that there's a enough overlap that this doesn't need to be a new widget.

Think about the following:
- What should this look like by default? I like the flat styling used in demo2 and a few other examples personally.
- Which elements should be configurable and how?
- What's the backwards compatible story?
- What are some possible nice looking pre-configured styles to use? Are there examples of tabs that we can draw from that would be nice to emulate?
- Do we actually need to skeuomorphize tabs (make them look like physical tabs), or does finding representations that are more digital work on the console (e.g. single line, use color / bold / etc. to highlight selection)
- What settings are required to support styling the individual parts of a tab (and what do we have curently)?
- What would be necessary to make the group of styles etc. configurable as a sort of theme type? (and more generally how would this apply to Ratatui's widgets?)

And then also consider if there's some things which are current problems for tabs
- wrapping / scrolling to select tabs at the end of a tab bar that goes off frame
- alignment of each tab (perhaps we should allow left / right alignment of individual tabs)?
- What other featues are missing that a tab bar needs?

If you want to iterate on this in a way that can be incorporated back into Ratatui easily, I'd encourage forking this widget (copy) into a small lib or within your own app, and then doing experiments on various parts incrementally. To make it easy to merge back into the core, focus a lot on backwards compatible changes when you can. Breaking changes are often fine when they significantly improve what's possible and don't make it hugely burdensome for existing code to upgrade.

You can also take the slower approach and work out all these things in a way and slowly apply them to the library, but I'd encourage experimentation to find what works well for your app and then bringing that knowledge to bear on the widget.

**Comment by crocuda** [comment]
Ok. 
So I'll make a separate widget.
Thank you for your promptitude and for pointing me out the pitfalls related to tabs.

**Comment by orhun** [comment]
Hey @pipelight!

I think we can keep it simple for the first step and add a new property to `Tabs` for this, e.g.

```rust
Tabs::new(vec!["Tab1", "Tab2", "Tab3", "Tab4"])
    .block(Block::bordered().title("Tabs"))
    .style(Style::default().white())
    .highlight_style(Style::default().yellow())
    .select(2)
    .divider(symbols::DOT)
    .tab_style(TabStyle::Glossy);  // <---
``` 

I would also recommend you to experiment with it on a separate widget first as well. Then we can incorporate those changes if there is enough overlap between `Tabs` and your implementation.

I think we can keep this issue open if you are planning to work on it :)

**Comment by joshka** [comment]
+1 to keeping this open, but perhaps we could just move it to discussions if there's no immediate action needed?
